src/sitemapper/builder.cr

```diff
@@ -24,36 +24,44 @@ module Sitemapper
     def generate : Array(Hash(String, String))
       paginator.total_pages.times do |page|
         filename = filename_for_page(page)
-        doc = XML.build(indent: " ", version: "1.0", encoding: "UTF-8") do |xml|
-          xml.element("urlset", xmlns: XMLNS_SCHEMA, "xmlns:video": XMLNS_VIDEO_SCHEMA, "xmlns:image": XMLNS_IMAGE_SCHEMA, "xmlns:xsi": XMLNS_XSI, "xsi:schemaLocation": XSI_SCHEMA_LOCATION) do
-            paginator.items(page + 1).each do |info|
-              build_xml_from_info(xml, info)
-            end
-          end
-        end
+        doc = build_xml_for_page(paginator.items(page + 1))
 
         @sitemaps << {"name" => filename, "data" => doc}
       end
 
       if @use_index
-        doc = XML.build(indent: " ", version: "1.0", encoding: "UTF-8") do |xml|
-          xml.element("sitemapindex", xmlns: XMLNS_SCHEMA, "xmlns:video": XMLNS_VIDEO_SCHEMA, "xmlns:image": XMLNS_IMAGE_SCHEMA, "xmlns:xsi": XMLNS_XSI, "xsi:schemaLocation": XSI_INDEX_SCHEMA_LOCATION) do
-            @sitemaps.each do |sm|
-              xml.element("sitemap") do
-                sitemap_name = sm["name"].to_s + (Sitemapper.config.compress ? ".gz" : "")
-                sitemap_url = [(Sitemapper.config.sitemap_host || @host), sitemap_name].join('/')
+        @sitemaps << generate_index
```

Comment:
this is refactored to allow for the streamer to subclass and reuse code to generate both a normal page and also the index